src/pip/_internal/download.py

```diff
@@ -199,13 +199,23 @@ def handle_401(self, resp, **kwargs):
 
         # Add our new username and password to the request
         req = HTTPBasicAuth(username or "", password or "")(resp.request)
+        req.register_hook("response", self.handle_401_again)
```

Comment:
Minor naming nit: we're not actually handling the 401, and definitely not doing so "again". Maybe `warn_on_401`?